File: geekbang/algorithm-models/2.3-差分代码模板/corpFlightBookings.js
// JavaScript
// LeetCode 1109 航班预订统计
/**
 * @param {number[][]} bookings
 * @param {number} n
 * @return {number[]}
 */
var corpFlightBookings = function(bookings, n) {
    var delta = new Array(n + 2).fill(0);  // 差分要开0~n+1
    for (const booking of bookings) {
        let fir = booking[0];
        let last = booking[1];
        let seats = booking[2];
        // 差分模板
        delta[fir] += seats;
        delta[last + 1] -= seats;
    }


    var a = new Array(n + 1).fill(0); // 0~n
    // 1~n
    for (let i = 1; i <= n; i++) a[i] = a[i - 1] + delta[i];
    // 0~n-1
    a.shift();
    return a;
};

File: geekbang/algorithm-models/2.3-差分代码模板/corpFlightBookings.py
# Python
# LeetCode 1109 航班预订统计
class Solution:
    def corpFlightBookings(self, bookings: List[List[int]], n: int) -> List[int]:
        delta = [0] * (n + 2);  # 差分要开0~n+1
        for booking in bookings:
            fir = booking[0]
            last = booking[1]
            seats = booking[2]
            # 差分模板
            delta[fir] += seats
            delta[last + 1] -= seats


        a = [0] * (n + 1)  # 0~n
        # 1~n
        for i in range(1, n + 1):
            a[i] = a[i - 1] + delta[i]
        # 0~n-1
        return a[1:]
